File: forms/united-kingdom-driver-and-vehicle-licensing-agency-m1-form/back-end-with-loco/src/engine/m1_validator.rs
use super::flagged_issues::detect_additional_flags;
use super::m1_rules::all_rules;
use super::types::{AssessmentData, FiredRule, ValidationResult};
use super::utils::{count_conditions, priority_order};

/// Pure function: validates a DVLA M1 assessment.
///
/// Conditional logic ported 1:1 from the front-end:
/// - Q1 = No  -> form stops at Q1; Q2/Q3 rules are skipped and the result
///               carries `stopped_at_q1 = true`.
/// - Q1 = Yes -> all rules apply.
pub fn validate_m1(data: &AssessmentData) -> ValidationResult {
    let stopped_at_q1 = data.diagnosis_confirmation.has_mental_health_diagnosis == "no";

    let mut fired_rules: Vec<FiredRule> = Vec::new();
    for rule in all_rules() {
        if stopped_at_q1 && (rule.id.starts_with("M1-Q2") || rule.id.starts_with("M1-Q3")) {
            continue;
        }
        if (rule.evaluate)(data) {
            fired_rules.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                priority: rule.priority.to_string(),
                description: rule.description.to_string(),
                message: rule.message.to_string(),
            });
        }
    }

    fired_rules.sort_by_key(|r| priority_order(&r.priority));

    let additional_flags = detect_additional_flags(data);
    let condition_count = count_conditions(data);

    let complete = fired_rules
        .iter()
        .filter(|r| r.category == "completeness")
        .count()
        == 0;

    ValidationResult {
        complete,
        stopped_at_q1,
        fired_rules,
        additional_flags,
        condition_count,
        timestamp: chrono::Utc::now().to_rfc3339(),
    }
}

File: forms/united-kingdom-driver-and-vehicle-licensing-agency-m1-form/front-end-with-html/js/m1-validator.js
import { detectAdditionalFlags } from './flagged-issues.js';
import { m1Rules } from './m1-rules.js';
import { priorityOrder } from './types.js';

// DVLA M1 form validator. Pure function: takes an `AssessmentData`, returns
// a `ValidationResult` with the fired rules, additional flags, completeness
// flag, stoppedAtQ1 flag, condition count, and a timestamp.
//
// Conditional logic:
//   - Q1 = No  -> form stops at Q1; downstream Q2/Q3 rules are skipped and
//                 the result is marked `stoppedAtQ1: true`.
//   - Q1 = Yes -> all rules apply.
//
// Ported 1:1 from front-end-form-with-svelte/src/lib/engine/m1-validator.ts.

/**
 * @typedef {import('./types.js').AssessmentData} AssessmentData
 * @typedef {import('./types.js').FiredRule} FiredRule
 * @typedef {import('./types.js').ValidationResult} ValidationResult
 */

/**
 * @param {AssessmentData} data
 * @returns {ValidationResult}
 */
function validateM1(data) {
  const stoppedAtQ1 =
    data.diagnosisConfirmation.hasMentalHealthDiagnosis === 'no';

  /** @type {FiredRule[]} */
  const firedRules = [];
  for (const rule of m1Rules) {
    // When the patient answered Q1 = No, skip Q2/Q3 rules entirely -
    // the DVLA instructions say not to complete those parts.
    if (
      stoppedAtQ1 &&
      (rule.id.startsWith('M1-Q2') || rule.id.startsWith('M1-Q3'))
    ) {
      continue;
    }
    try {
      if (rule.evaluate(data)) {
        firedRules.push({
          id: rule.id,
          category: rule.category,
          priority: rule.priority,
          description: rule.description,
          message: rule.message
        });
      }
    } catch (e) {
      console.warn(`M1 rule ${rule.id} evaluation failed:`, e);
    }
  }

  firedRules.sort(
    (a, b) => priorityOrder(a.priority) - priorityOrder(b.priority)
  );

  const additionalFlags = detectAdditionalFlags(data);
  const conditionCount = countConditions(data);

  const complete =
    firedRules.filter((r) => r.category === 'completeness').length === 0;

  return {
    complete,
    stoppedAtQ1,
    firedRules,
    additionalFlags,
    conditionCount,
    timestamp: new Date().toISOString()
  };
}

/** @param {AssessmentData} data */
function countConditions(data) {
  const c = data.mentalHealthConditions;
  let n = 0;
  if (c.anxietyDepressionWithoutImpairment === 'yes') n++;
  if (c.anxietyDepressionWithImpairment === 'yes') n++;
  if (c.bipolarAffectiveDisorder === 'yes') n++;
  if (c.eatingDisorder === 'yes') n++;
  if (c.ocdOrPtsd === 'yes') n++;
  if (c.personalityDisorder === 'yes') n++;
  if (c.schizophreniaOrPsychosis === 'yes') n++;
  if (c.other === 'yes') n++;
  return n;
}

export { validateM1, countConditions };
